File: rs_bindings_from_cc/test/bridging/string_test.rs
// Part of the Crubit project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

use cc_std;
use googletest::prelude::*;

#[gtest]
fn test_string_type_as_function_arg() {
    let s = cc_std::std::string::from("hello");
    let len = s.len();
    let ffi_len = string_test_lib::GetStringSize(s);
    assert_eq!(len, ffi_len.try_into().unwrap());
}

#[gtest]
fn test_string_type_as_return_value() {
    let s1: cc_std::std::string =
        unsafe { string_test_lib::CreateString("hello".as_ptr() as _, 5) };
    let s2 = cc_std::std::string::from("hello");
    assert_eq!(s1, s2);
}
// TODO(b/351976622): Support basic_string in supported.
// #[gtest]
// fn test_basic_string_as_return_value() {
//    let s1: cc_std::std::string =
//        unsafe { string_test_lib::CreateStringAsBasicString("hello".as_ptr()
// as _, 5) };    let s2 = cc_std::std::string::from("hello");
//    assert_eq!(s1, s2);
//}

File: cc_bindings_from_rs/test/bridging/string/string_test.cc
// Part of the Crubit project, under the Apache License v2.0 with LLVM
// Exceptions. See /LICENSE for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception

#include "cc_bindings_from_rs/test/bridging/string/string.h"

#include <string>

#include "gtest/gtest.h"

namespace crubit {
namespace {

TEST(StdStringBridging, Basic) {
  std::string s = "hello world";
  auto expected = string::roundtrip_string(s);
  static_assert(std::is_same_v<std::string, decltype(expected)>);
  EXPECT_EQ(expected, s);

  EXPECT_EQ(string::create_from_rust(), std::string("hello world from Rust"));
  EXPECT_EQ(string::compute_string_length("hello world"), 11);
}

}  // namespace
}  // namespace crubit